analyzers/src/SonarAnalyzer.CSharp/Rules/VariableShadowsField.cs

```diff
@@ -30,73 +30,79 @@ public sealed class VariableShadowsField : SonarDiagnosticAnalyzer
+                _ when VarPatternSyntaxWrapper.IsInstance(node) => ((VarPatternSyntaxWrapper)node).Designation,
+                _ when DeclarationPatternSyntaxWrapper.IsInstance(node) => ((DeclarationPatternSyntaxWrapper)node).Designation,
+                _ when ListPatternSyntaxWrapper.IsInstance(node) => ((ListPatternSyntaxWrapper)node).Designation,
+                _ => null
+            };
 
-        private static void ProcessVariableDesignation(SonarSyntaxNodeReportingContext context, VariableDesignationSyntaxWrapper variableDesignation)
+        private static void Process(SonarSyntaxNodeReportingContext context, SyntaxNode node)
         {
-            if (variableDesignation.AllVariables() is { Length: > 0 } variables
-                && context.ContainingSymbol.ContainingType is { } containingType
-                && GetMembers(containingType) is var members)
+            if (ExtractIdentifiers(node) is { Count: > 0 } identifiers
+                && GetContextSymbols(context) is var members)
             {
-                foreach (var variable in variables)
+                foreach (var identifier in identifiers)
                 {
-                    ReportOnVariableMatchingField(context, members, variable.Identifier);
+                    ReportOnVariableMatchingField(context, members, identifier);
                 }
             }
         }
 
-        private static void ProcessVariableDeclaration(SonarSyntaxNodeReportingContext context, VariableDeclarationSyntax variableDeclaration)
-        {
-            if (variableDeclaration is { Variables: { Count: > 0 } variables }
-                && context.ContainingSymbol.ContainingType is { } containingType
-                && GetMembers(containingType) is var members)
+        private static List<SyntaxToken> ExtractIdentifiers(SyntaxNode node) =>
+            node switch
             {
-                foreach (var variable in variables)
-                {
-                    ReportOnVariableMatchingField(context, members, variable.Identifier);
-                }
-            }
+                VariableDeclarationSyntax variableDeclaration => variableDeclaration.Variables.Select(x => x.Identifier).ToList(),
+                ForEachStatementSyntax foreachStatement => new() { foreachStatement.Identifier },
+                _ when VariableDesignationSyntaxWrapper.IsInstance(node) => ((VariableDesignationSyntaxWrapper)node).AllVariables().Select(x => x.Identifier).ToList(),
+                _ => new()
+            };
+
+        private static List<ISymbol> GetContextSymbols(SonarSyntaxNodeReportingContext context)
+        {
+            var members = context.ContainingSymbol.ContainingType.GetMembers();
+            var primaryConstructorParameters = members.FirstOrDefault(x => x.IsPrimaryConstructor())?.GetParameters();
+            var fieldsAndProperties = members.Where(x => x is IPropertySymbol or IFieldSymbol).ToList();
```

Comment:
I think we missed an [SemanticModel.IsAccessible](https://learn.microsoft.com/de-de/dotnet/api/microsoft.codeanalysis.semanticmodel.isaccessible?view=roslyn-dotnet-4.6.0) check here:
* Private fields/properties are accessible from within the declaring type but not derived types
* protected/public fields/properties are accessible from the declaring and derived types.

Just add a comment about this in #8260